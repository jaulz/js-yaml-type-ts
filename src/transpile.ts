import { CompilerOptions } from 'typescript'
import ts from 'typescript'
import dump from './dump'

export default function transpile(
  code: string | object,
  compilerOptions: CompilerOptions = {}
): string {
  const result = ts.transpileModule(
    typeof code === 'string' ? code : `Object.assign(exports, ${dump(code)})`,
    {
      reportDiagnostics: true,
      compilerOptions: {
        module: ts.ModuleKind.CommonJS,
        target: ts.ScriptTarget.ES5,
        jsx: ts.JsxEmit.React,
        noEmitHelpers: false,
        sourceMap: false,
        inlineSourceMap: false,
        ...compilerOptions,
      },
    }
  )

  // Check if there was a compilation error
  const diagnostics = result.diagnostics
  if (diagnostics && diagnostics.length > 0) {
    const message = diagnostics[0].messageText
    const error = typeof message === 'string' ? message : message.messageText

    throw new Error(`SyntaxError: ${error}`)
  }

  return result.outputText.trim()
}
