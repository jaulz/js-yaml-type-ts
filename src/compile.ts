import transpile from './transpile'
import { VMScript, NodeVM, NodeVMOptions } from 'vm2'
import { CompilerOptions } from 'typescript'
import { OriginalCodeSymbol, TranspiledCodeSymbol } from '.'

export default function compile<T = any>(
  code: string,
  compilerOptions?: CompilerOptions,
  vmOptions?: NodeVMOptions
): T {
  // Create VM
  const vm = new NodeVM({
    timeout: 10000,
    sandbox: {},
    ...vmOptions,
  })

  return vm.run(new VMScript(code))
}
