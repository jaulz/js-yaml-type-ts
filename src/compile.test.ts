import compile from './compile'
import transpile from './transpile'

describe('compile', () => {
  it('compile code correctly', () => {
    const content = compile(
      transpile(`
      export default {
        boolean: true,
        func: () => true,
        asyncFunc: async () => true,
        jsx: (React) => <Test />,
      }
    `)
    )

    expect(content).not.toBeNull()
    expect(typeof content).toEqual('object')
    expect(content).toMatchSnapshot()
  })
})
