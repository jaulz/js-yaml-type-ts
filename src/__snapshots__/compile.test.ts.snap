// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile compile code correctly 1`] = `
Object {
  "default": Object {
    "asyncFunc": [Function],
    "boolean": true,
    "func": [Function],
    "jsx": [Function],
  },
}
`;
